sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,655 @@
+{
+  // 'I' is the compiler time constant information being annotated on 'a'
+  [[special_annotations::use_case_x_attribute(I)]] SomeType* a;
+  ...
+  *a = some_val;
+  ...
+  read_val = *a
+}
+```
+However, this mechanism fails to meet requirements (2) and (3) listed above
+since:
+
+  * Adding new attributes can add non-trivial attribute creation overheads for
+  compiler developers. With growing set of attributes, there is an associated
+  burden of maintaining them and of preparing diagnostic checks.
+  * Compilers are not required to preserve information propagated as attributes.
+  Hence this method is only suitable for hints, and not functional directives.
+  Further, from a given load/store call a compiler may not be able to
+  successfully trace back to the declaration to deduce which annotation was
+  applied to it.
+
+The `annotated_ptr` class described in this document is a class template that
+encapsulates a pointer. The template accepts a list of compile-time constant
+properties. The implementation of the class defined here should preserve the
+information provided as compile-time constant properties through all uses of the
+pointer unless noted otherwise.
+
+
+=== Representation of `annotated_ptr`
+
+`annotated_ptr` is a class template, parameterized by the type of the underlying
+allocation `T`, and a list of properties `PropertyListT`.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ptr {
+  ...
+----
+
+`PropertyListT` enables properties to be associated with an `annotated_ptr`.
+Properties may be specified for an `annotated_ptr` to provide semantic
+modification or optimization hint information.
+
+Here's an example of how a property could be used:
+
+[source,c++]
+----
+using namespace sycl::ext::oneapi::experimental;
+{
+  sycl::queue q;
+  // A pointer to an integer
+  auto ptr = ...
+  // alignment of the pointer in bytes specified using the property 'alignment'
+  auto arg_a = annotated_ptr(ptr, properties{alignment<4>});
+
+  q.submit([=]{
+    ...
+    *arg_a = (*arg_a) * 2;
+  });
+  ...
+}
+----
+
+The section below and the table that follows, describe the constructors, member
+functions and factory methods for `annotated_ptr`.
+The section below refers to an `annotated_ref` class which is described in the
+section following this one.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ptr {
+  public:
+    using reference = annotated_ref<T, PropertyListT>;
+
+    annotated_ptr() noexcept;
+    explicit annotated_ptr(T *Ptr, const properties<PropertyListT> &) noexcept;
+    annotated_ptr(annotated_ptr const &) noexcept;
+    template <T, typename P> explicit annotated_ptr(
+      annotated_ptr<T, P> const &) noexcept;
+    template <T, typename PropertyListU, typename PropertyListV>
+    explicit annotated_ptr(annotated_ptr<T, PropertyListU> const &,
```

Comment:
Can you provide a deduction guide that allows users to use this constructor with "auto"?

Reply:
Brox is working on this.